<?xml version="1.0"?>
<robot name="ghost1" xmlns:xacro="http://ros.org/wiki/xacro">

    <xacro:include filename="$(find ghost_sim)/urdf/macros/actuator_macros.xacro"/>
    <xacro:include filename="$(find ghost_sim)/urdf/macros/sensor_macros.xacro"/>
    
    <!-- Import Primary Robot URDF -->
    <xacro:include filename="$(find ghost_description)/urdf/ghost1.xacro"/>

    <!-- Add Sensors -->
    <!-- <xacro:lidar_macro link_name="lidar_link" topic_name="scan"></xacro:lidar_macro> -->
    <!-- <xacro:imu_macro parent_link="camera_view_link"></xacro:imu_macro> -->

    <!-- Joint State Publisher -->
    <gazebo>
        <plugin name="v5_robot_plugin" filename="libv5_robot_plugin.so">
            <joint_names>
                steering_joint_left
                steering_joint_right
                steering_joint_back
                steering_top_test
                wheel_joint_left
                wheel_joint_right
                wheel_joint_back
                wheel_top_test
            </joint_names>
            <!-- Motor names must be formatted in this manner in order to index actuator jacobian -->
            <motor_names>
                DRIVE_LEFT_FRONT_LEFT_MOTOR
                DRIVE_LEFT_FRONT_RIGHT_MOTOR
                DRIVE_LEFT_BACK_LEFT_MOTOR
                DRIVE_LEFT_BACK_RIGHT_MOTOR
                DRIVE_RIGHT_FRONT_LEFT_MOTOR
                DRIVE_RIGHT_FRONT_RIGHT_MOTOR
                DRIVE_RIGHT_BACK_LEFT_MOTOR
                DRIVE_RIGHT_BACK_RIGHT_MOTOR
            </motor_names>

            <encoder_names>
                STEERING_FRONT_LEFT_ENCODER
                STEERING_FRONT_RIGHT_ENCODER
                STEERING_BACK_LEFT_ENCODER
                STEERING_BACK_RIGHT_ENCODER
            </encoder_names>

            <!-- System Jacobians -->
                <!-- To describe the relationship between motors/encoders and robot joints in a universal way, we encode them -->
                <!-- as matrices which relate the vector of motor/encoder speeds to joint speeds. We can also take a pseudo-inverse -->
                <!-- to relate torques. Examples are included below. -->

                <!-- Example 1 -->
                    <!--         motor_1 motor_2    -->
                    <!-- joint_1    1       0       -->
                    <!-- joint_2    0       1       -->
        
                    <!-- Both joints are directly driven by their respective motors, and their velocites and torques match. -->

                <!-- Example 2 -->
                <!--          motor_1   motor_2  motor_3    -->
                    <!-- joint_1    0.5       0.5       0       -->
                    <!-- joint_2     0         0        2       -->

                    <!-- joint_1 is driven by motor_1 and motor_2. It moves at the average of the two motor velocites, and with -->
                    <!-- the torque of both motors added together. This represents two motors attached to the same system --> 
                    <!-- joint_2 moves twice as fast as motor_3, representing an external gear reduction on the motor -->

                <!-- In general, describe the way the velocities are related and the torque relationship will be extracted inside the plugin -->

            <!-- This matrix describes the relationship between motor velocities and joint velocities -->
            <!-- Width = # Motors, Height = # Joints -->
            <actuator_jacobian>
              0    0    0    0    0    0    0    0
              0    0    0    0    0    0    0    0 
              0    0    0    0    0    0    0    0
            1.0    0    0    0    0    0    0    0
              0    0    0    0    0    0    0    0
              0    0    0    0    0    0    0    0
              0    0    0    0    0    0    0    0
              0    0    0    0    0    0    0    0
            </actuator_jacobian>

            <!-- This matrix describes the relationship between the encoders and the joints -->
            <!-- This encoder matrix does not include encoders within motors. The relation ship between motor encoders and 
            joints is defined in the actuator_jacobian -->
            <!-- Width = # Encoders, Height = # Joints -->
            <sensor_jacobian>
            1 0 0 0
            0 1 0 0
            0 0 1 0
            0 0 0 1
            0 0 0 0
            0 0 0 0
            0 0 0 0
            0 0 0 0
            </sensor_jacobian>
        </plugin>
    </gazebo>
</robot>